tsdb/tsm1/engine.go

```diff
@@ -1104,15 +1116,66 @@ func (e *Engine) compactFull(ctx context.Context, grp CompactionGroup, wg *sync.
 
 	// Try the lo priority limiter, otherwise steal a little from the high priority if we can.
 	if e.compactionLimiter.TryTake() {
+		// Attempt to get ownership of the semaphore for this engine. If the
+		// default semaphore is in use then ownership will always be granted.
+		ttl := influxdb.DefaultLeaseTTL
+		lastCompaction := time.Duration(atomic.LoadInt64(&e.lastFullCompactionDuration))
+		if lastCompaction > ttl {
+			ttl = lastCompaction // If the last full compaction took > default ttl then set a new TTL
+		}
+
+		lease, err := e.fullCompactionSemaphore.TryAcquireTTL(ctx, ttl)
+		if err != nil {
+			e.logger.Warn("Failed to execute full compaction", zap.Error(err), zap.Duration("semaphore_requested_ttl", ttl))
+			e.compactionLimiter.Release()
+			return false
+		} else if lease == nil {
+			e.logger.Info("Cannot acquire semaphore ownership to carry out full compaction", zap.Duration("semaphore_requested_ttl", ttl))
+			e.compactionLimiter.Release()
+			return false
+		}
+
+		done := make(chan struct{}) // Closed when compaction finished.
+		go func(lease influxdb.Lease) {
+			ttl, err := lease.TTL(ctx)
+			if err != nil {
+				e.logger.Warn("unable to get TTL for lease on semaphore", zap.Error(err))
+				ttl = influxdb.DefaultLeaseTTL // This is probably a reasonable fallback.
+			}
+
+			// Renew the lease when ttl is halved
+			ticker := time.NewTicker(ttl / 2)
+			for {
+				select {
+				case <-done:
+					if err := lease.Release(ctx); err != nil {
+						e.logger.Warn("Lease on sempahore was not released", zap.Error(err))
+					}
+					return
+				case <-ticker.C:
+					if err := lease.KeepAlive(ctx); err != nil {
+						e.logger.Warn("Unable to extend lease", zap.Error(err))
+					} else {
+						e.logger.Info("Extended lease on semaphore")
+					}
+				}
+			}
+		}(lease)
```

Comment:
This method is already pretty long. Can you move this goroutine to a separate method?